### ISSUE TITLE
grpc-ruby 1.42.0 fails with dyld[]: missing symbol called

### ISSUE THREAD

**Issue by NivedhaSenthil** [grpc/grpc#28271]
### What version of gRPC and what language are you using?
grpc - 1.42.0
language - ruby

### What operating system (Linux, Windows,...) and version?
MacOS - 12.0.1
MacBook Pro (16-inch, 2019)
Processor 2.6 GHz 6-Core Intel Core i7

### What runtime / compiler are you using (e.g. python version or version of gcc)
Ruby - 2.7.0, 3.0.0, 3.0.3

### What did you do?
Run acceptance for the spanner ruby client library.
https://github.com/googleapis/google-cloud-ruby/tree/main/google-cloud-spanner
```
# some random test project and key, just so that test starts
export SPANNER_TEST_PROJECT="test"
export SPANNER_TEST_KEYFILE=~/test.json
bundle exec rake acceptance
```

### What did you expect to see?

All test runs fine without any crash by GRPC

### What did you see instead?

Fails while making an rpc call. The same works fine when we downgrade to 1.41.1.

```
D1203 18:48:33.343635000 4538004992 ev_posix.cc:172]                   Using polling engine: poll
D1203 18:48:33.343708000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "grpclb"
D1203 18:48:33.343717000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "priority_experimental"
D1203 18:48:33.343724000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "weighted_target_experimental"
D1203 18:48:33.343730000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "pick_first"
D1203 18:48:33.343734000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "round_robin"
D1203 18:48:33.343740000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "ring_hash_experimental"
D1203 18:48:33.343746000 4538004992 dns_resolver_ares.cc:495]          Using ares dns resolver
D1203 18:48:33.343784000 4538004992 certificate_provider_registry.cc:33] registering certificate provider factory for "file_watcher"
D1203 18:48:33.343789000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "cds_experimental"
D1203 18:48:33.343795000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "xds_cluster_impl_experimental"
D1203 18:48:33.343803000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "xds_cluster_resolver_experimental"
D1203 18:48:33.343808000 4538004992 lb_policy_registry.cc:42]          registering LB policy factory for "xds_cluster_manager_experimental"
D1203 18:48:33.343882000 4538004992 rb_grpc.c:291]                     GRPC_RUBY: grpc_ruby_init - prev g_grpc_ruby_init_count:0
D1203 18:48:33.343906000 4538004992 rb_grpc.c:291]                     GRPC_RUBY: grpc_ruby_init - prev g_grpc_ruby_init_count:1
D1203 18:48:33.343916000 4538004992 rb_grpc.c:291]                     GRPC_RUBY: grpc_ruby_init - prev g_grpc_ruby_init_count:2
D1203 18:48:33.343922000 4538004992 rb_grpc.c:291]                     GRPC_RUBY: grpc_ruby_init - prev g_grpc_ruby_init_count:3
D1203 18:48:33.343931000 4538004992 rb_channel.c:708]                  GRPC_RUBY: wait for channel polling thread to start
D1203 18:48:33.343940000 123145478193152 rb_grpc.c:291]                GRPC_RUBY: grpc_ruby_init - prev g_grpc_ruby_init_count:4
D1203 18:48:33.343968000 123145481912320 rb_channel.c:697]             GRPC_RUBY: run_poll_channels_loop - create connection polling thread
D1203 18:48:33.343977000 123145481912320 rb_grpc.c:291]                GRPC_RUBY: grpc_ruby_init - prev g_grpc_ruby_init_count:5
D1203 18:48:33.343982000 123145481912320 rb_channel.c:621]             GRPC_RUBY: run_poll_channels_loop_no_gil - begin
dyld[60801]: missing symbol called
rake aborted!
SignalException: SIGABRT
```

**Comment by quartzmo** [comment]
I am also able to reproduce this:

```sh
Using googleapis-common-protos-types 1.3.0
Using grpc 1.42.0 (x86_64-darwin) (was 1.41.1)
Using googleapis-common-protos 1.3.12
Using gapic-common 0.7.0
Using google-cloud-firestore-v1 0.4.5 from source at `../google-cloud-firestore-v1`
Using google-cloud-firestore 2.6.5 from source at `.`
Bundle updated!
quartzmo@Chriss-iMac google-cloud-firestore % bundle exec rake acceptance
Run options: --seed 35205

# Running:

dyld: lazy symbol binding failed: Symbol not found: ___isOSVersionAtLeast
  Referenced from: /Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/grpc-1.42.0-x86_64-darwin/src/ruby/lib/grpc/3.0/grpc_c.bundle
  Expected in: flat namespace

dyld: Symbol not found: ___isOSVersionAtLeast
  Referenced from: /Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/grpc-1.42.0-x86_64-darwin/src/ruby/lib/grpc/3.0/grpc_c.bundle
  Expected in: flat namespace

rake aborted!
SignalException: SIGABRT
/Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-firestore/Rakefile:70:in `block in <top (required)>'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/cli/exec.rb:63:in `load'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/cli/exec.rb:63:in `kernel_load'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/cli/exec.rb:28:in `run'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/cli.rb:494:in `exec'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/cli.rb:30:in `dispatch'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/cli.rb:24:in `start'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/exe/bundle:49:in `block in <top (required)>'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/lib/bundler/friendly_errors.rb:130:in `with_friendly_errors'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/bundler-2.2.16/exe/bundle:37:in `<top (required)>'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/bin/bundle:23:in `load'
/Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/bin/bundle:23:in `<main>'
Tasks: TOP => acceptance:run
(See full trace by running task with --trace)
quartzmo@Chriss-iMac google-cloud-firestore % sw_vers
ProductName:	macOS
ProductVersion:	11.6
BuildVersion:	20G165
quartzmo@Chriss-iMac google-cloud-firestore % ruby -v
ruby 3.0.1p64 (2021-04-05 revision 0fb782ee38) [x86_64-darwin19]
```

**Comment by mr-salty** [comment]
I'm also seeing this:

```
dyld: lazy symbol binding failed: Symbol not found: ___isOSVersionAtLeast
  Referenced from: /Users/todd/.rbenv/versions/2.6.9/lib/ruby/gems/2.6.0/gems/grpc-1.42.0-x86_64-darwin/src/ruby/lib/grpc/2.6/grpc_c.bundle
  Expected in: flat namespace
  
$ sw_vers; ruby -v
ProductName:	macOS
ProductVersion:	11.6.1
BuildVersion:	20G224
ruby 2.6.9p207 (2021-11-24 revision 67954) [x86_64-darwin20]
```

**Comment by mr-salty** [comment]
(and can confirm downgrading to 1.41 fixes it)

**Comment by apolcyn** [comment]
The mac package build was revised in 1.42.0, and this is very likely related: we didn't publish the old `universal-darwin` gem and instead published the `x86_64-darwin` gem as the binary package for mac.

The universal-darwin gem is still present in the 1.42.0 pre-release, though, in https://rubygems.org/gems/grpc/versions/1.42.0.pre1-universal-darwin as another potential workaround.

I'm not able to reproduce this easily right now.

**Comment by quartzmo** [comment]
Confirming that `1.42.0.pre1 (x86_64-darwin)` reproduces:

```
Using grpc 1.42.0.pre1 (x86_64-darwin) (was 1.41.1)
Using minitest-focus 1.3.1
Using googleapis-common-protos 1.3.12
Using googleauth 1.1.0
Using google-cloud-core 1.6.0 from source at `../google-cloud-core`
Using google-style 1.25.2
Using gapic-common 0.7.0
Using google-cloud-firestore-v1 0.4.5 from source at `../google-cloud-firestore-v1`
Using google-cloud-firestore 2.6.5 from source at `.`
Bundle updated!
quartzmo@Chriss-iMac google-cloud-firestore % bundle exec rake acceptance
Run options: --seed 22827

# Running:

dyld: lazy symbol binding failed: Symbol not found: ___isOSVersionAtLeast
  Referenced from: /Users/quartzmo/.rbenv/versions/3.0.1/gemsets/quartzmo/gems/grpc-1.42.0.pre1-x86_64-darwin/src/ruby/lib/grpc/3.0/grpc_c.bundle
  Expected in: flat namespace
```

**Comment by apolcyn** [comment]
@quartzmo can you please check if `1.42.0.pre1 (universal-darwin)` works? (that's the gem that uses the old build)

**Comment by quartzmo** [comment]
Yes, sorry that took a little more effort, but I just confirmed that it works:

```
quartzmo@Chriss-iMac google-cloud-firestore % gem install grpc-1.42.0.pre1-universal-darwin.gem
Successfully installed grpc-1.42.0.pre1-universal-darwin
Parsing documentation for grpc-1.42.0.pre1-universal-darwin
Installing ri documentation for grpc-1.42.0.pre1-universal-darwin
Done installing documentation for grpc after 0 seconds
1 gem installed
quartzmo@Chriss-iMac google-cloud-firestore % gem uninstall grpc                               

Select gem to uninstall:
 1. grpc-1.42.0.pre1-universal-darwin
 2. grpc-1.42.0.pre1-x86_64-darwin
 3. All versions
> 2
Successfully uninstalled grpc-1.42.0.pre1-x86_64-darwin
quartzmo@Chriss-iMac google-cloud-firestore % b rake acceptance
Run options: --seed 56477

# Running:

.

Finished in 0.918545s, 1.0887 runs/s, 20.6849 assertions/s.

1 runs, 19 assertions, 0 failures, 0 errors, 0 skips

```

**Comment by apolcyn** [comment]
OK, that seems to narrow things down to the new x86_64 package build from https://github.com/grpc/grpc/pull/25794.

cc @flavorjones

**Comment by flavorjones** [comment]
I'm able to reproduce this with both the spanner and firestore projects. Digging in.

**Comment by flavorjones** [comment]
A good explanation of what's happening is in this series of blog posts:

1. [Undefined Symbol on macOS – Eugene Petrenko](https://jonnyzzz.com/blog/2018/05/16/link-error/)
2. [Undefined isOSVersionAtLeast on macOS – Eugene Petrenko](https://jonnyzzz.com/blog/2018/06/05/link-error-2/)
3. [Understanding isOSVersionAtLeast on macOS – Eugene Petrenko](https://jonnyzzz.com/blog/2018/06/13/link-error-3/)

And some prior instances of this exact problem (cross-compiling via osxcross):

- [A few issues building Qt · Issue #187 · tpoechtrager/osxcross](https://github.com/tpoechtrager/osxcross/issues/187)
- [___isOSVersionAtLeast undefined on OSX · Issue #1325 · ddnet/ddnet](https://github.com/ddnet/ddnet/issues/1325)

It doesn't look like the rake-compiler-dock installation of osxcross has the compiler-rt libraries installed. I'm going to explore doing that next.

**Comment by flavorjones** [comment]
Confirmed, building the gem in a rake-compiler-dock image container that provides `compiler-rt` resolves this symbol and the repros noted above no longer fail on a `dyld` "symbol not found" error.

@apolcyn I'm going to submit a PR today to rake-compiler-dock that will address this; and then the grpc project will need to pull in that change to its RCD container-builder process.

**Comment by flavorjones** [comment]
@apolcyn see https://github.com/rake-compiler/rake-compiler-dock/pull/60 for upstream progress.

If you'd like to cherry-pick the fix, here it is: https://github.com/rake-compiler/rake-compiler-dock/pull/60/commits/655f8acae058ac7a17b6b550f7b1bb15f1da0c9d

**Comment by dazuma** [comment]
@apolcyn I'm not sure when the upstream rake-compiler fix is going to be in. Is it possible to release a 1.42.1 (or even a 1.42.0.1) reinstating the universal-darwin build instead?

**Comment by jnarowski** [comment]
@flavorjones if I fork this repo and cherry pick your commit, will I be able to build and run GRPC on an m1 mac? Not sure if I need to build any binaries for it, or if just forking and adding that above code would be enough?

**Comment by flavorjones** [comment]
@jnarowski I'm not sure if you're asking "how do I create a precompiled native gem for arm64-darwin" or if you're asking "does grpc install on M1".

If you simply want to run grpc on your M1, you can install 1.42.0 and it should compile correctly for you during installation:

```text
$ uname -a
Darwin soho.local 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000 arm64

$ ruby -v
ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [arm64-darwin21]

$ ruby -e 'puts Gem::Platform.local'
arm64-darwin-21

$ gem install grpc
Fetching grpc-1.42.0.gem
Building native extensions. This could take a while...
Successfully installed grpc-1.42.0
1 gem installed
```

If you're asking "how do I create a precompiled native gem for arm64-darwin" I can provide steps, but it's a little bit more involved and I'd like to wait for confirmation from you that that's what you want before digging in.

**Comment by jnarowski** [comment]
I just want this gem to run on my M1 mac. Don't need to precompile a native gem. 

I am on `ruby 2.6.6` installed via `rbenv`. I tried installing `1.42.0` and that didn't work for me. 

```
$ uname -a
Darwin Wolverine.local 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000 arm64

$ ruby -v
ruby 2.6.6p146 (2020-03-31 revision 67876) [-darwin21]

$ ruby -e 'puts Gem::Platform.local'
-darwin-21

$ gem install grpc
Fetching grpc 1.42.0
Installing grpc 1.42.0 with native extensions
```

When I try running something that depends on this gem, I get the following error:

```
LoadError: dlopen(/Users/jnarowski/.rbenv/versions/2.6.6/lib/ruby/gems/2.6.0/gems/grpc-1.42.0/src/ruby/lib/grpc/grpc_c.bundle, 0x0009): symbol not found in flat namespace '_gpr_asprintf' - /Users/jnarowski/.rbenv/versions/2.6.6/lib/ruby/gems/2.6.0/gems/grpc-1.42.0/src/ruby/lib/grpc/grpc_c.bundle
from /Users/jnarowski/.rbenv/versions/2.6.6/lib/ruby/gems/2.6.0/gems/activesupport-5.0.7.1/lib/active_support/dependencies.rb:293:in `require'
```

**Comment by jonmchan** [comment]
I'm confirming seeing this issue on the following macos system:

```
$ uname -a
Darwin Jonathans-MacBook-Pro.local 19.6.0 Darwin Kernel Version 19.6.0: Tue Oct 12 18:34:05 PDT 2021; root:xnu-6153.141.43~1/RELEASE_X86_64 x86_64

$ ruby -v
ruby 2.7.5p203 (2021-11-24 revision f69aeb8314) [x86_64-darwin19]

$ ruby -e 'puts Gem::Platform.local'
x86_64-darwin-19

```


Error when trying to initiate a gRPC client call:

```
dyld: lazy symbol binding failed: Symbol not found: ___isOSVersionAtLeast
  Referenced from: /Users/jonathan/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/grpc-1.42.0-x86_64-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle
  Expected in: flat namespace

dyld: Symbol not found: ___isOSVersionAtLeast
  Referenced from: /Users/jonathan/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/grpc-1.42.0-x86_64-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle
  Expected in: flat namespace

Abort trap: 6
```

Ran it in linux in a docker instance and it worked without issue.

```
$ docker run --network="fabric_test" -it -v "$(pwd)":/app/ ruby:2.7.5  /bin/bash
```

**Comment by michaelschmitz** [comment]
Same issue as @jonmchan also just tried a simple call, can install 1.42.0 without problems, but still fails with the dyld error:
```
$client.document_text_detection image: file_name
dyld: lazy symbol binding failed: Symbol not found: ___isOSVersionAtLeast
Referenced from: /Users/ms/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/grpc-1.42.0-x86_64-darwin/src/ruby/lib/grpc/2.5/grpc_c.bundle

$ uname -a
Darwin MSs-MacBook-Pro.fritz.box 20.6.0 Darwin Kernel Version 20.6.0: Wed Jun 23 00:26:31 PDT 2021; root:xnu-7195.141.2~5/RELEASE_X86_64 x86_64

$ruby -v
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-darwin17]

$ ruby -e 'puts Gem::Platform.local'
x86_64-darwin-17

```

If the solution is to precompile the gem natively, then I would appreciate your advice @flavorjones.

**Comment by flavorjones** [comment]
@michaelschmitz Yes, what you're describing is the bug reported in this issue. My advice above to the grpc maintainers requires them to update their build environment and ship a new precompiled gem. In the meantime as a workaround you can install the v1.42.0 vanilla ruby gem and compile it upon installation.

@jnarowski Your issue looks different, please open a separate issue? Thanks.

**Comment by michaelschmitz** [comment]
@flavorjones Really appreciate your advice!

For now - because I just needed to solve one operational issue that was blocking a team - I simply wrote my test code and pushed it to an Heroku instance where I could then play with the remote console to fine tune my results, unblock the team and move to the next operational task.

**Comment by doctorpangloss** [comment]
`abseil-cpp` needs to be patched to resolve these issues easily. Then use `lipo` to merge the two architectures together.

**Comment by KidA001** [comment]
Just for anyone else who is looking for a (temporary) solve, downgrading `grcp` to `1.41.1` in my gemfile fixed my issue. I was trying to use the `google-cloud-firestore` gem and it was failing due to the error listed in this issue.

**Comment by brodock** [comment]
to anyone who needs to get `grpc` working on their development machines, you can temporarily set bundler `platform` to `ruby` which ignores any OS specific pre-built binary and rebuild directly from source:

```
bundle config set force_ruby_platform true
```

**Comment by lukelove** [comment]
thank you @brodock 🙏

**Comment by lucianokrebs** [comment]
@brodock's suggestion did not work for me. Instead, what did the trick for me was adding this into my `Gemfile`:

```rb
if RUBY_PLATFORM =~ /darwin/
  module BundlerHack
    def __materialize__
      if name == 'grpc' || name == 'google-protobuf'
        Bundler.settings.temporary(force_ruby_platform: true) do
          super
        end
      else
        super
      end
    end
  end
  Bundler::LazySpecification.prepend(BundlerHack)
end
```

Considerations:
- I'm only using it for development.
- `ruby -e 'puts Gem::Platform.local'` returns `x86_64-darwin-20` for me
- I'm not using an M1 MacBook pro

**Comment by flavorjones** [comment]
@apolcyn I think this can be closed now that #29304 has been merged.
